protocol/contact.go

```diff
@@ -383,3 +403,56 @@ func (c *Contact) MarshalJSON() ([]byte, error) {
 
 	return json.Marshal(item)
 }
+
+// ContactRequestPropagatedStateReceived handles the propagation of state from
+// the other end.
+func (c *Contact) ContactRequestPropagatedStateReceived(state *protobuf.ContactRequestPropagatedState) ContactRequestProcessingResponse {
+
+	// It's inverted, as their local states is our remote state
+	expectedLocalState := ContactRequestState(state.RemoteState)
+	expectedLocalClock := state.RemoteClock
+
+	remoteState := ContactRequestState(state.LocalState)
+	remoteClock := state.LocalClock
+
+	response := ContactRequestProcessingResponse{}
+
+	// If we notice that the state is not consistent, and their clock is
+	// outdated, we send back the state so they can catch up.
+	if expectedLocalClock < c.ContactRequestLocalClock && expectedLocalState != c.ContactRequestLocalState {
+		response.processed = true
+		response.sendBackState = true
+	}
+
+	// If they expect our state to be more up-to-date, we only
+	// trust it if the state is set to None, in this case we can trust
+	// it, since a retraction can be initiated by both parties
+	if expectedLocalClock > c.ContactRequestLocalClock && c.ContactRequestLocalState != ContactRequestStateDismissed && expectedLocalState == ContactRequestStateNone {
+		response.processed = true
+		c.ContactRequestLocalClock = expectedLocalClock
+		c.ContactRequestLocalState = ContactRequestStateNone
+		// We set they remote state, as this was an implicit retraction
```

Comment:
So, it means that we don't know whether the contact request was retracted expliciltly, for example by receiving a `ContactRequestRetractedMessage`, but we noticed that their state has changed so that effectively the contact request flow was restarted, this could have been done on their end, and we haven't (yet) received the retraction message, or on our end from a paired device, and we have not received that message.
I'll update the comment to make it clearer

Reply:
Thank you for the clarification :bulb: